### ISSUE TITLE
Failure to Send Media on Reconnection after WebSocket Disconnect

### ISSUE THREAD

**Issue by ynigun** [amigniter/mod_audio_stream#29]
## Description

After an unexpected disconnection from the WebSocket server, the `uuid_audio_stream` command successfully reconnects but fails to send media data, although it can still send text.

## Current Behavior

1. Initial connection works correctly, sending both text and media.
2. When the remote WebSocket server disconnects unexpectedly:
   - Subsequent `uuid_audio_stream $UUID start` successfully connects to the WebSocket server.
   - The new connection can send text data.
   - The new connection fails to send any media data.

## Expected Behavior

After an unexpected WebSocket server disconnection:
1. `uuid_audio_stream $UUID start` should fully reconnect, sending both text and media data.

## Steps to Reproduce

1. Start a FreeSWITCH session and connect to a WebSocket server using the following command:
   ```
   uuid_audio_stream $UUID start ws://$WS_SERVER:$PORT/ws mixed 8k mix11.wav
   ```
2. Verify that the initial connection is working correctly (sending both text and media).
3. Simulate a server disconnection (e.g., stop the WebSocket server or network interruption).
4. Attempt to reconnect using the same command as in step 1.
5. Observe that the connection is reestablished and text can be sent, but no media is transmitted.

## Environment

- FreeSWITCH version: 1.10.11

## Logs

**Note: The following logs are from custom debug statements added to the code and are not part of the original library.**

Here are the relevant logs from the reconnection attempt:

```
2024-06-24 15:51:49.051812 99.67% [DEBUG] mod_audio_stream.c:161 mod_audio_stream cmd: $UUID start ws://$WS_SERVER:$PORT/ws mixed 8k mix11.wav
2024-06-24 15:51:49.051812 99.67% [DEBUG] mod_audio_stream.c:82 calling stream_session_init.
2024-06-24 15:51:49.051812 99.67% [DEBUG] audio_streamer_glue.cpp:310 Entering stream_data_init
2024-06-24 15:51:49.051812 99.67% [DEBUG] audio_streamer_glue.cpp:322 ($UUID) Initializing: desired sampling: 8000, channels: 1, rtp_packets: 1
2024-06-24 15:51:49.051812 99.67% [DEBUG] audio_streamer_glue.cpp:328 ($UUID) Initial metadata set: mix11.wav
2024-06-24 15:51:49.051812 99.67% [DEBUG] audio_streamer_glue.cpp:333 ($UUID) Calculated buffer length: 320
2024-06-24 15:51:49.051812 99.67% [DEBUG] audio_streamer_glue.cpp:336 ($UUID) Creating AudioStreamer with wsUri: ws://$WS_SERVER:$PORT/ws
2024-06-24 15:51:49.051812 99.67% [DEBUG] audio_streamer_glue.cpp:341 ($UUID) AudioStreamer created and assigned to tech_pvt->pAudioStreamer: 0x7f3ce4077400
2024-06-24 15:51:49.051812 99.67% [DEBUG] audio_streamer_glue.cpp:345 ($UUID) Mutex initialized
2024-06-24 15:51:49.051812 99.67% [DEBUG] audio_streamer_glue.cpp:360 $UUID: initializing buffer(320) to adjusted 512 bytes
2024-06-24 15:51:49.051812 99.67% [DEBUG] audio_streamer_glue.cpp:388 ($UUID) no resampling needed for this call
2024-06-24 15:51:49.051812 99.67% [DEBUG] audio_streamer_glue.cpp:391 ($UUID) stream_data_init
2024-06-24 15:51:49.051812 99.67% [DEBUG] mod_audio_stream.c:89 Adding media bug with flags: 3
2024-06-24 15:51:49.051812 99.67% [DEBUG] mod_audio_stream.c:90 Session address: 0x7f3ce8039038, UserData: 0x7f3ce41543f8
2024-06-24 15:51:49.051812 99.67% [DEBUG] mod_audio_stream.c:28 capture_callback: INIT
2024-06-24 15:51:49.051812 99.67% [DEBUG] switch_core_media_bug.c:976 Attaching BUG to sofia/external/$DID@$IP_ADDRESS
2024-06-24 15:51:49.051812 99.67% [DEBUG] mod_audio_stream.c:97 Media bug added successfully. Bug address: 0x7f3ce416b138
2024-06-24 15:51:49.051812 99.67% [DEBUG] mod_audio_stream.c:98 Setting private data on channel. Channel: 0x7f3ce8037110
2024-06-24 15:51:49.051812 99.67% [DEBUG] mod_audio_stream.c:103 Retrieved bug from channel private data. Address: 0x7f3ce416b138
2024-06-24 15:51:49.051812 99.67% [DEBUG] mod_audio_stream.c:105 Exiting start_capture. Status: 0
2024-06-24 15:51:49.191822 99.67% [DEBUG] audio_streamer_glue.cpp:131 sending initial metadata mix11.wav
```

After the last log entry ("sending initial metadata mix11.wav"), no further logs are generated and the process appears to be stuck at this point.

*Note: This issue was drafted with the assistance of claude.ai :)*

**Comment by amigniter** [comment]
Thanks for your report. I'm sure with all the feedback from the community this little module is getting close to bug free state.
Already addressed this. Until I commit a fix you can try applying this change and let me know the results.
In mod_audio_stream.c , line 33 change it to:
```C
case SWITCH_ABC_TYPE_CLOSE:
            {
                switch_log_printf(SWITCH_CHANNEL_SESSION_LOG(session), SWITCH_LOG_INFO, "Got SWITCH_ABC_TYPE_CLOSE.\n");
                stream_session_cleanup(session, NULL, 0);
            }
            break;
```

Let me know if you are interested to know, and I will explain what's happening, what is this change doing and why.

**Comment by ynigun** [comment]
Thank you for the quick response and the suggested fix. I applied the change, and I'm pleased to report that it resolved the issue. The module is now able to reconnect and stream media properly after a disconnection from the server.

From what I can tell, it seems that this modification is affecting the removal of the media bug by changing how the `channelIsClosing` parameter is passed to `stream_session_cleanup`.

I'm particularly interested in understanding the implications of this change:

1. Are there any potential side effects I should be aware of?
2. Could this change lead to any memory leaks or resource management issues?
3. What was happening before this change that caused the reconnection problem?

Thank you so much for your help and quick response. I really appreciate your time and expertise!

**Comment by amigniter** [comment]
@ynigun,

Thanks for your report. Right, as expected, it should now reconnect and send audio again during the same session.

> From what I can tell, it seems that this modification is affecting the removal of the media bug by changing how the channelIsClosing parameter is passed to stream_session_cleanup.

Exactly! I am happy to explain what is happening under the hood, especially when the user is interested to know.
You got it right, it is about removing the attached bug. When channelIsClosing, parameter 1 passed to cleanup function, media bug won't be removed (but private data associated with it is null'ed and everything we allocated is cleared). The idea is to leave it to FreeSWITCH to remove the bug (since the channel is closing, meaning the session is ending) we will do the cleanup (so no potential leaks, all cleaned up here) but leave the bug to FS's internal handlers to remove it (and it's taking care of everything, FreeSWITCH is really awesome). It is aware of the bug added to the channel, bug is closed already, and it will remove it together with everything else associated to the session when it is cleared. Here is the catch, the channel is not closing in this case (websocket got disconnected but the channel is still active). If you follow the handlers(events) inside the websocket methods, on `close` event I am calling media bug close. Second parameter is whether to destroy it or not and we just want to close the bug and not to destroy it. What it is doing is sending the signal to the bug and it enters the closing state, `SWITCH_ABC_TYPE_CLOSE`, and we're calling the cleanup here. But, it is not all over yet as we can see now. Channel is not closing, websocket got disconnected and we want to restart the streaming. That's why we have to remove the bug (not to leave it to FreeSWITCH to remove it), so it can be re-attached. Reason for it is to make sure `start` is called with an attached bug. You called the API and the streaming started; you call it again (and we're not adding a new bug, it is already attached and the streaming is running, so nothing happens and you're given an error). In case of an error (websocket disconnected) we will clean everything and remove the bug so on the next `start` it can be re-attached to the channel. Now, one question arises: can we reuse the bug (if it wasn't removed on still active channel) ? I think it will shed even more light to this. Yes we could, but we would need to use one more api command like `restart` besides the `start` but I don't want to make it more complicated than this. It is much easier to completely remove it so we can use `start` command on restart and also on fresh new start. There are no leaks, everything is cleanup properly and there are no side effects. You are absolutely safe. The only thing, which I explained now, is the bug removal which was left to FreeSWITCH to remove it, and now we are taking care of it.
